### PR TITLE
Add check for forbidden item_id in item properties (RAD-720)

### PR DESCRIPTION
**Pull request by foglcz**

### CODE REVIEW

File: src/Model/Command/ItemProperty.php [hunk 1]
```diff
@@ -17,7 +17,13 @@ class ItemProperty extends AbstractCommand
     private function __construct(string $itemId, array $properties)
     {
         $this->setItemId($itemId);
+
         $this->properties = $properties;
+        Assertion::keyNotExists(
```
Comment (OndraM) [inline src/Model/Command/ItemProperty.php]: I'd create protected setter for properties an put the assertion inside.

**Review by OndraM — CHANGES_REQUESTED** [review]

**Comment by OndraM** [comment]
And what is the reason for this? Its because the item property name is there by default always?

And the item_id cannot be updated once item is created?

Maybe the reasons should be also explained in the exception message.

**Comment by foglcz** [comment]
+ add chagnelog

**Comment by foglcz** [comment]
Right. So, `item_id` is something like a primary key. In original tests:

```php
$ip = ItemProperty::create('exampleItemId', ['item_id' => 'someOtherItemId']);
dump($ip->getCommandParameters());
// ['item_id' => 'examleItemId']
```

While correct behaviour (property should not overwrite the id of the item), I felt that this would be confusing to the developer.

File: tests/Model/Command/ItemPropertyTest.php [hunk 2]
```diff
@@ -24,6 +25,15 @@ public function shouldBeInstantiableViaNamedConstructor(array $properties, array
         );
     }
 
+    /**
+     * @test
+     */
+    public function shouldNotAllowItemIdInProperties(): void
```
Comment (OndraM) [inline tests/Model/Command/ItemPropertyTest.php]: This should be placed after the following data provider

File: src/Model/Command/ItemProperty.php [hunk 3]
```diff
@@ -32,6 +32,17 @@ protected function setItemId(string $itemId): void
         $this->itemId = $itemId;
     }
 
+    protected function setProperties(array $properties): void
+    {
+        Assertion::keyNotExists(
+            $properties,
+            'item_id',
+            '$properties cannot contain key "item_id" - please use different key name.'
```
Comment (OndraM) [inline src/Model/Command/ItemProperty.php]: I'm not sure "please use different key name" is the best advice.

Maybe we should rather explain the item_id is already used for identifying the item and it cannot be updated.

File: src/Model/Command/ItemPropertySetup.php [hunk 4]
```diff
@@ -60,6 +60,7 @@ public static function set(string $propertyName): self
     protected function setPropertyName(string $propertyName): void
     {
         Assertion::typeIdentifier($propertyName);
+        Assertion::notEq($propertyName, 'item_id', 'Cannot define property with name item_id - please use different name');
```
Comment (OndraM) [inline src/Model/Command/ItemPropertySetup.php]: Dtto, maybe instead explain the item_id is set up in Matej by default and don't need to be set up explicitly.

**Review by OndraM — CHANGES_REQUESTED** [review]

File: tests/Model/Command/ItemPropertyTest.php [hunk 5]
```diff
@@ -2,10 +2,20 @@
...
 use PHPUnit\Framework\TestCase;
 
 class ItemPropertyTest extends TestCase
 {
+    /**
+     * @test
+     */
+    public function shouldNotAllowItemIdInProperties(): void
+    {
+        $this->expectException(DomainException::class);
```
Comment (OndraM) [inline tests/Model/Command/ItemPropertyTest.php:15]: I'd add expectExceptionMessage - to make sure the exception occurred in the expected case.

File: src/Model/Command/ItemPropertySetup.php [hunk 6]
```diff
@@ -60,6 +60,11 @@ public static function set(string $propertyName): self
     protected function setPropertyName(string $propertyName): void
     {
         Assertion::typeIdentifier($propertyName);
+        Assertion::notEq(
+            $propertyName,
+            'item_id',
+            'Cannot redefine property item_id - it is used as primary key to identify items. Please use different name.'
```
Comment (OndraM) [inline src/Model/Command/ItemPropertySetup.php]: `Cannot manipulate with property "item_id" - it is used by Matej to identify items.` or something?

File: src/Model/Command/ItemProperty.php [hunk 7]
```diff
@@ -32,6 +32,17 @@ protected function setItemId(string $itemId): void
         $this->itemId = $itemId;
     }
 
+    protected function setProperties(array $properties): void
+    {
+        Assertion::keyNotExists(
+            $properties,
+            'item_id',
+            '$properties cannot contain key item_id - it is used as primary key to identify the item. Please use different key instead.'
```
Comment (OndraM) [inline src/Model/Command/ItemProperty.php]: `Cannot update value of "item_id" property - it is used by Matej to identify the item and cannot be altered once created.` ?

**Review by OndraM — CHANGES_REQUESTED** [review]

**Review by OndraM — APPROVED** [review]

**Pull request merged**